dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -882,6 +902,266 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+                        + " outputstream for DOI zip file for upload to Zenodo."
+                        + " Error is " + fne.getMessage(), HttpStatus.SC_INTERNAL_SERVER_ERROR);
+            }
+            writeStreamAsZip(sourceFiles, outputStream, path);
+            java.nio.file.Path zipPath = Paths.get(fileName);
+            File zipFile = zipPath.toFile();
+
+            try {
+                DepositionFile file = filesApi.createFile(depositionID, zipFile, fileName);
+            } catch (ApiException e) {
+                LOG.error("Could not create files for new version on Zenodo. Error is " + e.getMessage());
+                throw new CustomWebApplicationException("Could not create files for new version on Zenodo."
+                        + " Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+            }
+        }
+    }
+
+    /**
+     * Register a Zenodo DOI for the workflow version
+     * @param workflow    workflow for which DOI is registered
+     * @param workflowVersion workflow version for which DOI is registered
+     */
+    private void registerZenodoDOIForWorkflow(Workflow workflow, WorkflowVersion workflowVersion) {
+        ApiClient zendoClient = new ApiClient();
+        // for testing, either 'https://sandbox.zenodo.org/api' or 'https://zenodo.org/api' is the first parameter
+        String zenodoUrlApi = zenodoUrl + "/api";
+        zendoClient.setBasePath(zenodoUrlApi);
+        String zenodoAccessToken = System.getProperty("ZENODO_TOKEN");
+        zendoClient.setApiKey(zenodoAccessToken);
+
+        DepositsApi depositApi = new DepositsApi(zendoClient);
+        ActionsApi actionsApi = new ActionsApi(zendoClient);
+
+        Deposit deposit = new Deposit();
+        Deposit returnDeposit;
+        String workflowVersionDoiURL = workflowVersion.getDoiURL();
+        if (workflowVersionDoiURL != null && !workflowVersionDoiURL.isEmpty()) {
+            LOG.error("Workflow version " + workflowVersion.getName() + " already has DOI " + workflowVersionDoiURL
+                    + ". Dockstore will only create one DOI per version.");
+            throw new CustomWebApplicationException("Workflow version " + workflowVersion.getName() + " already has DOI "
+                    + workflowVersionDoiURL + ". Dockstore will only create one DOI per version.", HttpStatus.SC_METHOD_NOT_ALLOWED);
+        }
+
+        // Find out if this workflow already has at least one
+        // version that has been assigned a DOI
+        // If a version DOI exists, we will create another version DOI
+        // instead of creating a new workflow concept DOI and version DOI
+        // Get the ID of one of the workflow version DOIs
+        // because Zenodo requires that we use it to create the next
+        // workflow version DOI
+        String latestWorkflowVersionDOIURL = null;
+        Set<WorkflowVersion> setOfWorkflowVersions = workflow.getWorkflowVersions();
```

Comment:
Ditto here, easier to read as a stream. Can be done in a follow-up PR.

Reply:
next sprint https://ucsc-cgl.atlassian.net/browse/DOCK-831

Reply:
next sprint https://ucsc-cgl.atlassian.net/browse/DOCK-831